### PR TITLE
Fixes for API kind documentation

### PR DESCRIPTION
**Pull request by ajreynol**
Also renames str.tolower -> str.to_lower, str.toupper -> str.to_upper.

### CODE REVIEW

File: src/api/cpp/cvc5.h [hunk 1]
```diff
@@ -4281,6 +4281,9 @@ class CVC5_EXPORT Solver
   /**
    * Declare a symbolic pool of terms with the given initial value.
    *
+   * For details on how pools are used to specify instructions for quantifier
+   * instantiation, see documentation for the INST_POOL kind.
```
Comment (aniemetz) [inline src/api/cpp/cvc5.h:4285]:
```suggestion
   * instantiation, see documentation for the #INST_POOL kind.
```
Can you check if this correctly links to the `INST_POOL` kind with this fix?

File: src/api/cpp/cvc5_kind.h [hunk 2]
```diff
@@ -342,20 +342,11 @@ enum Kind : int32_t
...
+   * uinterpreted Sort @f$S@f$ is less than or equal to an upper bound.
    *
+   * - Arity: `0`
    * - Create Term of this Kind with:
-   *   - Solver::mkTerm(Kind, const std::vector<Term>&) const
-   *   - Solver::mkTerm(const Op&, const std::vector<Term>&) const
-   *
-   * - Create Op of this kind with:
-   *   - Solver::mkOp(Kind) const
+   *   - Solver::mkCardinalityConstraint(const Sort& sort, uint32_t upperBound) const
```
Comment (aniemetz) [inline src/api/cpp/cvc5_kind.h:349]:
```suggestion
   *   - Solver::mkCardinalityConstraint(const Sort&, uint32_t) const
```

File: src/api/cpp/cvc5_kind.h [hunk 3]
```diff
@@ -2213,7 +2204,8 @@ enum Kind : int32_t
    */
   APPLY_UPDATER,
   /**
-   * Match expression.
+   * Match expression. This purpose of this kind is primarily to support
+   * the SMT-LIBv2 match expression.
```
Comment (aniemetz) [inline src/api/cpp/cvc5_kind.h:2208]:
```suggestion
   * Match expression.
   
   * This kind is primarily used in the parser to support the
   * SMT-LIBv2 `match` expression.
```

File: src/api/cpp/cvc5_kind.h [hunk 4]
```diff
@@ -4040,7 +4032,7 @@ enum Kind : int32_t
    * Instantiation pattern.
    *
    * Specifies a (list of) terms to be used as a pattern for quantifier
-   * instantiation.
+   * instantiation. Should only be used as a child of INST_PATTERN_LIST.
```
Comment (aniemetz) [inline src/api/cpp/cvc5_kind.h:4035]:
```suggestion
   * instantiation.
   *
   * @note Should only be used as a child of #INST_PATTERN_LIST.
```

File: src/api/cpp/cvc5_kind.h [hunk 5]
```diff
@@ -4057,7 +4049,7 @@ enum Kind : int32_t
    * Instantiation no-pattern.
    *
    * Specifies a (list of) terms that should not be used as a pattern for
-   * quantifier instantiation.
+   * quantifier instantiation. Should only be used as a child of INST_PATTERN_LIST.
```
Comment (aniemetz) [inline src/api/cpp/cvc5_kind.h:4052]:
```suggestion
   * quantifier instantiation.
   *
   * @note Should only be used as a child of #INST_PATTERN_LIST.
```

File: src/api/cpp/cvc5_kind.h [hunk 6]
```diff
@@ -4071,10 +4063,29 @@ enum Kind : int32_t
    */
   INST_NO_PATTERN,
   /**
-   * Instantiation pool.
+   * Instantiation pool annotation.
+   *
+   * Specifies an annotation for pool based instantiation.  Should only be used
+   * as a child of INST_PATTERN_LIST.
```
Comment (aniemetz) [inline src/api/cpp/cvc5_kind.h:4069]:
```suggestion
   * Specifies an annotation for pool based instantiation.
   *
   * Should only be used as a child of #INST_PATTERN_LIST.
```

File: src/api/cpp/cvc5_kind.h [hunk 7]
```diff
@@ -4071,10 +4063,29 @@ enum Kind : int32_t
...
    *
-   * Specifies an annotation for pool based instantiation.
+   * In detail, pool symbols can be declared via the method
+   *  - Solver::declarePool(const std::string&, const Sort&, const std::vector<Term>&) const
```
Comment (aniemetz) [inline src/api/cpp/cvc5_kind.h:4072]: I think we need an empty line after this for it to render correctly. Can you check?

File: src/api/cpp/cvc5_kind.h [hunk 8]
```diff
@@ -4088,10 +4099,29 @@ enum Kind : int32_t
    */
   INST_POOL,
   /**
-   * A instantantiation-add-to-pool annotation.
+   * A instantantiation-add-to-pool annotation. Should only be used as a child
+   * of INST_PATTERN_LIST.
```
Comment (aniemetz) [inline src/api/cpp/cvc5_kind.h:4103]:
```suggestion
   * A instantantiation-add-to-pool annotation.
   *
   * @note Should only be used as a child of #INST_PATTERN_LIST.
```

File: src/api/cpp/cvc5_kind.h [hunk 9]
```diff
@@ -4088,10 +4099,29 @@ enum Kind : int32_t
...
+   * \rst
+   * .. code:: lisp
    *
-   * - Arity: `1`
-   *   - `1:` The pool to add to.
+   *     (FORALL (VARIABLE_LIST x) F 
+   *             (INST_PATTERN_LIST (INST_ADD_TO_POOL (ADD x 1) p)))
+   * \endrst
+   * 
+   * where assume that x has type Int. When this quantified formula is
```
Comment (aniemetz) [inline src/api/cpp/cvc5_kind.h:4118]:
```suggestion
   * where assume that @f$x@f$ has type Int. When this quantified formula is
```

File: src/api/cpp/cvc5_kind.h [hunk 10]
```diff
@@ -4088,10 +4099,29 @@ enum Kind : int32_t
...
+   * instantiated with e.g. the term t, the term (ADD t 1) is added to pool p.
```
Comment (aniemetz) [inline src/api/cpp/cvc5_kind.h:4119]:
```suggestion
   * instantiated with, e.g., the term @f$t@f$, the term `(ADD t 1)` is added to pool @f$p@f$.
```

File: src/api/cpp/cvc5_kind.h [hunk 11]
```diff
@@ -4088,10 +4099,29 @@ enum Kind : int32_t
...
+   *
+   * - Arity: `2`
+   *   - `1:` The Term whose free variables are bound by the quantified formula.
+   *   - `2:` The pool to add to, whose type should be a set of elements that
+   *          match the sort of the first argument.
```
Comment (aniemetz) [inline src/api/cpp/cvc5_kind.h:4124]: Does this have set sort then?

File: src/api/cpp/cvc5_kind.h [hunk 12]
```diff
@@ -4088,10 +4099,29 @@ enum Kind : int32_t
...
+   *     (FORALL (VARIABLE_LIST x) F 
+   *             (INST_PATTERN_LIST (INST_ADD_TO_POOL (ADD x 1) p)))
+   * \endrst
+   * 
+   * where assume that x has type Int. When this quantified formula is
+   * instantiated with e.g. the term t, the term (ADD t 1) is added to pool p.
+   *
+   * - Arity: `2`
+   *   - `1:` The Term whose free variables are bound by the quantified formula.
+   *   - `2:` The pool to add to, whose type should be a set of elements that
```
Comment (aniemetz) [inline src/api/cpp/cvc5_kind.h:4123]:
```suggestion
   *   - `2:` The pool to add to, whose Sort should be a set of elements that
```

File: src/api/cpp/cvc5_kind.h (hunk 11 above)
Comment (aniemetz) [inline src/api/cpp/cvc5_kind.h:4124]:
```suggestion
   *          match the Sort of the first argument.
```

File: src/api/cpp/cvc5_kind.h [hunk 13]
```diff
@@ -4102,10 +4132,30 @@ enum Kind : int32_t
    */
   INST_ADD_TO_POOL,
   /**
-   * A skolemization-add-to-pool annotation.
+   * A skolemization-add-to-pool annotation.  Should only be used as a child of
+   * INST_PATTERN_LIST.
```
Comment (aniemetz) [inline src/api/cpp/cvc5_kind.h:4136]:
```suggestion
   * A skolemization-add-to-pool annotation.
   *
   * Should only be used as a child of #INST_PATTERN_LIST.
```

File: src/api/cpp/cvc5_kind.h [hunk 14]
```diff
@@ -4102,10 +4132,30 @@ enum Kind : int32_t
...
    *
-   * - Arity: `1`
-   *   - `1:` The pool to add to.
+   *     (FORALL (VARIABLE_LIST x) F 
+   *             (INST_PATTERN_LIST (SKOLEM_ADD_TO_POOL (ADD x 1) p)))
+   * \endrst
+   * 
+   * where assume that x has type Int. When this quantified formula is
+   * skolemized, e.g. with k of type Int, then the term (ADD k 1) is added to
+   * the pool p.
```
Comment (aniemetz) [inline src/api/cpp/cvc5_kind.h:4153]:
```suggestion
   * where assume that @f$x@f$ has type Int. When this quantified formula is
   * skolemized, e.g., with @f$k@f$ of type Int, then the term `(ADD k 1)` is
   * added to the pool @f$p@f$.
```

File: src/api/cpp/cvc5_kind.h [hunk 15]
```diff
@@ -4102,10 +4132,30 @@ enum Kind : int32_t
...
+   *
+   * - Arity: `2`
+   *   - `1:` The Term whose free variables are bound by the quantified formula.
+   *   - `2:` The pool to add to, whose type should be a set of elements that
+   *          match the sort of the first argument.
```
Comment (aniemetz) [inline src/api/cpp/cvc5_kind.h:4158]:
```suggestion
   *   - `2:` The pool to add to, whose Sort should be a set of elements that
   *          match the Sort of the first argument.
```

Comment (aniemetz) [reply src/api/cpp/cvc5_kind.h:4158]: And ditto with my question above.

File: src/api/cpp/cvc5_kind.h [hunk 16]
```diff
@@ -4119,7 +4169,8 @@ enum Kind : int32_t
    * Instantiation attribute.
    *
    * Specifies a custom property for a quantified formula given by a
-   * term that is ascribed a user attribute.
+   * term that is ascribed a user attribute. Should only be used as a child of
+   * INST_PATTERN_LIST.
```
Comment (aniemetz) [inline src/api/cpp/cvc5_kind.h:4173]:
```suggestion
   * term that is ascribed a user attribute.
   *
   * @note Should only be used as a child of #INST_PATTERN_LIST.
```

**Review by aniemetz — CHANGES_REQUESTED** [review]

File: src/api/cpp/cvc5_kind.h (hunk 11 above)
Comment (ajreynol) [reply src/api/cpp/cvc5_kind.h]: Yes.

File: src/api/cpp/cvc5_kind.h [hunk 17]
```diff
@@ -4014,10 +4012,30 @@ enum Kind : int32_t
...
+   * \rst
+   * .. code:: lisp
+   *
+   *     (FORALL (VARIABLE_LIST x y) F (INST_PATTERN_LIST (INST_POOL p q)))
+   * \endrst
+   * 
+   * if x and y have types T1 and T2, then pool symbols p and q should have
+   * types (Set T1) and (Set T2) respectively. This annotation specifies that
+   * the quantified formula above should be instantiated with the product of
+   * all terms that occur in the sets p and q.
```
Comment (aniemetz) [inline src/api/cpp/cvc5_kind.h:4037]:
```suggestion
   * if @f$x@f$ and @f$y@f$ have Sorts @f$S_1@f$ and @f$S_2@f$,
   * then pool symbols @f$p@f$ and @f$q@f$ should have Sorts
   * (Set @f$S_1@f$) and (Set @f$S_2@f$), respectively.
   * This annotation specifies that the quantified formula above should be
   * instantiated with the product of all terms that occur in the sets @f$p@f$
   * and @f$q@f$.
```

File: src/api/cpp/cvc5_kind.h (hunk 7 above)
Comment (aniemetz) [reply src/api/cpp/cvc5_kind.h:4022]: Yep, we do: 
https://cvc5.github.io/docs-ci/docs-pr8397/api/cpp/kind.html#_CPPv4N4cvc53api4Kind9INST_POOLE

**Review by aniemetz — APPROVED** [review]

**Pull request merged**